examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,1762 @@
+                image, text = batch
+
+                image = image.to(accelerator.device, non_blocking=True)
+                encoded_text = compute_embeddings_fn(text)
+
+                pixel_values = image.to(dtype=weight_dtype)
+                if vae.dtype != weight_dtype:
+                    vae.to(dtype=weight_dtype)
+
+                # encode pixel values with batch size of at most 32
+                latents = []
+                for i in range(0, pixel_values.shape[0], 32):
+                    latents.append(vae.encode(pixel_values[i : i + 32]).latent_dist.sample())
+                latents = torch.cat(latents, dim=0)
+
+                latents = latents * vae.config.scaling_factor
+                latents = latents.to(weight_dtype)
+                bsz = latents.shape[0]
+
+                # 2. Sample random student timesteps s uniformly in `student_timestep_schedule` and sample random
+                # teacher timesteps t uniformly in [0, ..., noise_scheduler.config.num_train_timesteps - 1].
+                student_index = torch.randint(0, student_distillation_steps, (bsz,), device=latents.device).long()
+                student_timesteps = student_timestep_schedule[student_index]
+                teacher_timesteps = torch.randint(
+                    0, noise_scheduler.config.num_train_timesteps, (bsz,), device=latents.device
+                ).long()
+
+                # 3. Sample noise and add it to the latents according to the noise magnitude at each student timestep
+                # (that is, run the forward process on the student model)
+                student_noise = torch.randn_like(latents)
+                noisy_student_input = noise_scheduler.add_noise(latents, student_noise, student_timesteps)
+
+                # 4. Prepare prompt embeds and unet_added_conditions
+                prompt_embeds = encoded_text.pop("prompt_embeds")
+
+                # 5. Get the student model predicted original sample `student_x_0`.
+                student_noise_pred = unet(
+                    noisy_student_input,
+                    student_timesteps,
+                    encoder_hidden_states=prompt_embeds.float(),
+                ).sample
+                student_x_0 = denoiser(student_noise_pred, student_timesteps, noisy_student_input)
+
+                # 6. Sample noise and add it to the student's predicted original sample according to the noise
+                # magnitude at each teacher timestep (that is, run the forward process on the teacher model, but
+                # using `student_x_0` instead of latents sampled from the prior).
+                teacher_noise = torch.randn_like(student_x_0)
+                noisy_teacher_input = noise_scheduler.add_noise(student_x_0, teacher_noise, teacher_timesteps)
+
+                # 7. Sample random guidance scales w ~ U[w_min, w_max] for CFG.
+                w = (args.w_max - args.w_min) * torch.rand((bsz,)) + args.w_min
```

Comment:
It might be better to use fixed cfg scale. That's what they do in `sds` right ?

Reply:
In that case, we could just set `w_min` and `w_max` to same value. So this is fine.

Reply:
> That's what they do in `sds` right ?

At least with respect to the [DreamFusion paper](https://arxiv.org/pdf/2209.14988.pdf), it's not obvious how they handled the CFG scale: I believe they don't mention it in Appendices A.2 and A.3, where they discuss their training setup.

My thinking is that using a range of guidance scales $`\omega`$ would allow the student model to generalize to a wide range of guidance scales and makes using CFG with different guidance scales when sampling from the distilled student make more sense. In particular setting $`\omega_{min} = 1`$ (using the Imagen CFG formulation $`\hat{\epsilon}_\theta(\boldsymbol{z}_t, \boldsymbol{c}, t) = \omega \cdot \epsilon_\theta(\boldsymbol{z}_t, \boldsymbol{c}, t) + (1 - \omega) \cdot \epsilon_\theta(\boldsymbol{z}_t, \boldsymbol{\emptyset}, t)`$) allows the student to learn the conditional teacher output $`\epsilon_\theta(\boldsymbol{z}_t, \boldsymbol{c}, t)`$, which could be useful for applications like guidance rescaling.